kernel/kernel-api/src/test/scala/io/delta/kernel/internal/actions/AddFileSuite.scala

```diff
@@ -0,0 +1,234 @@
+      path = "test/path",
+      partitionValues = Map("a" -> "1"),
+      size = 1L,
+      modificationTime = 10L,
+      dataChange = false,
+      deletionVector = Option.empty,
+      tags = Option(Map("tag1" -> "value1")),
+      baseRowId = Option(30L),
+      defaultRowCommitVersion = Option(40L),
+      stats = Option("{\"numRecords\":100}")
+    )
+
+    val addFile = new AddFile(addFileRow)
+    assert(addFile.getPath === "test/path")
+    assert(VectorUtils.toJavaMap(addFile.getPartitionValues).asScala.equals(Map("a" -> "1")))
+    assert(addFile.getSize === 1L)
+    assert(addFile.getModificationTime === 10L)
+    assert(addFile.getDataChange === false)
+    assert(addFile.getDeletionVector === Optional.empty())
+    assert(VectorUtils.toJavaMap(addFile.getTags.get()).asScala.equals(Map("tag1" -> "value1")))
+    assert(addFile.getBaseRowId === Optional.of(30L))
+    assert(addFile.getDefaultRowCommitVersion === Optional.of(40L))
+    // DataFileStatistics doesn't have an equals() override, so we need to compare the string
+    assert(addFile.getStats.get().toString === "{\"numRecords\":100}")
+    assert(addFile.getNumRecords === Optional.of(100L))
+  }
+
+  test("update a single field of an AddFile") {
+    val addFileRow = generateTestAddFileRow(baseRowId = Option(1L))
+    val addFileAction = new AddFile(addFileRow)
+
+    val updatedAddFileAction = addFileAction.withNewBaseRowId(2L)
+    assert(updatedAddFileAction.getBaseRowId === Optional.of(2L))
+
+    val updatedAddFileRow = updatedAddFileAction.toRow
+    assert(new AddFile(updatedAddFileRow).getBaseRowId === Optional.of(2L))
+  }
+
+  test("update multiple fields of an AddFile multiple times") {
+    val baseAddFileRow =
+      generateTestAddFileRow(
+        path = "test/path",
+        baseRowId = Option(0L),
+        defaultRowCommitVersion = Option(0L)
+      )
+    var addFileAction = new AddFile(baseAddFileRow)
+
+    (1L until 10L).foreach { i =>
+      addFileAction = addFileAction
+        .withNewBaseRowId(i)
+        .withNewDefaultRowCommitVersion(i * 10)
+
+      assert(addFileAction.getPath === "test/path")
+      assert(addFileAction.getBaseRowId === Optional.of(i))
+      assert(addFileAction.getDefaultRowCommitVersion === Optional.of(i * 10))
+    }
+  }
+
+  test("toString() prints all fields of AddFile with partitionValues / tags in sorted order") {
+    val addFileRow = generateTestAddFileRow(
+      path = "test/path",
+      partitionValues = Map("b" -> "2", "a" -> "1", "c" -> "3"),
+      size = 100L,
+      modificationTime = 1234L,
+      dataChange = false,
+      tags = Option(Map("tag1" -> "value1", "tag2" -> "value2")),
+      baseRowId = Option(12345L),
+      defaultRowCommitVersion = Option(67890L),
+      stats = Option("{\"numRecords\":10000}")
+    )
+    val addFile = new AddFile(addFileRow)
+    val expectedString = "AddFile{" +
+      "path='test/path', " +
+      "partitionValues={a=1, b=2, c=3}, " +
```

Comment:
ah, I see the confusion here. We want these printed in the order they are defined in the schema (this is how they will appear in the file system, too)